crates/flycheck/src/lib.rs

```diff
@@ -326,13 +319,43 @@ impl CargoActor {
 
             read_at_least_one_message = true;
 
-            // Skip certain kinds of messages to only spend time on what's useful
-            match &message {
-                cargo_metadata::Message::CompilerArtifact(artifact) if artifact.fresh => (),
-                cargo_metadata::Message::BuildScriptExecuted(_) => (),
-                _ => self.sender.send(message).unwrap(),
+            // Try to deserialize a message from Cargo.
+            {
+                let mut deserializer = serde_json::Deserializer::from_str(&message);
+                deserializer.disable_recursion_limit();
+                if let Ok(message) = cargo_metadata::Message::deserialize(&mut deserializer) {
+                    // Skip certain kinds of messages to only spend time on what's useful
+                    match message {
+                        cargo_metadata::Message::CompilerArtifact(artifact) if !artifact.fresh => {
+                            self.sender.send(CargoMessage::CompilerArtifact(artifact)).unwrap()
+                        }
+                        cargo_metadata::Message::CompilerMessage(msg) => {
+                            self.sender.send(CargoMessage::Diagnostic(msg.message)).unwrap()
+                        }
+
+                        cargo_metadata::Message::CompilerArtifact(_)
+                        | cargo_metadata::Message::BuildScriptExecuted(_)
+                        | cargo_metadata::Message::BuildFinished(_)
+                        | cargo_metadata::Message::TextLine(_)
+                        | _ => (),
+                    }
+                }
+            }
+
+            // Try to deserialize a Diagnostic directly from Rustc.
+            {
+                let mut deserializer = serde_json::Deserializer::from_str(&message);
+                deserializer.disable_recursion_limit();
+                if let Ok(message) = Diagnostic::deserialize(&mut deserializer) {
+                    self.sender.send(CargoMessage::Diagnostic(message)).unwrap()
+                }
             }
         }
         Ok(read_at_least_one_message)
     }
 }
+
```

Comment:
```suggestion
#[derive(Serde)] #[serde(untagged)]
```

I *think* this will allow you to avoid manually writing that if-else.

Reply:
Seems to work.  I made a new enum `JsonMessage` that covers that, so that most of the Cargo types are still filtered out before sending over the channel.